### PR TITLE
Stop rendering dead links

### LINKED ISSUE #560: Dead links to hidden items appearing in documentation
In the `Field` module in base we have the following code:
```ocaml
(**/**)
module For_generated_code : sig
  (*_ don't use this by hand, it is only meant for ppx_fields_conv *)

  type ('perm, 'record, 'field) t =
    { force_variance : 'perm -> unit
    ; name : string
    ; setter : ('record -> 'field -> unit) option
    ; getter : 'record -> 'field
    ; fset : 'record -> 'field -> 'record
    }

  val opaque_identity : 'a -> 'a
end
(**/**)

(** ['record] is the type of the record.  ['field] is the type of the
    values stored in the record field with name [name]. ['perm] is a way
    of restricting the operations that can be used. *)
type ('perm, 'record, 'field) t_with_perm =
  | Field of ('perm, 'record, 'field) For_generated_code.t
[@@unboxed]
```

The `For_generated_code` is correctly left out of the documentation, however the `t_with_perm` type is documented as:
```
type ('perm, 'record, 'field) t_with_perm =
  | Field of ('perm, 'record, 'field) For_generated_code.t
```
with `For_generated_code.t` a dead link. It should instead be documented as:
```
type ('perm, 'record, 'field) t_with_perm
```

### PR DESCRIPTION
**Pull request by lubegasimon**
fixes #560

### CODE REVIEW

File: test/html/test.ml [hunk 1]
```diff
@@ -289,6 +289,9 @@ let source_files_post408 =
 
 let source_files_pre410 = [ ("bugs_pre_410.ml", [ "Bugs_pre_410/index.html" ]) ]
 
+let source_files_stop_dead_link_doc =
```
Comment (Julow) [inline test/html/test.ml:292]: The convention would be to call this `source_files_post404`. This way more tests can be added to this list.

File: src/model/names.ml [hunk 2]
```diff
@@ -31,12 +48,12 @@ module Name : Name = struct
 
   let to_string = function
     | Std s -> s
-    | Internal (s, i) -> Printf.sprintf "$%s$%d" s i
+    | Internal (s, i) -> Printf.sprintf "{%s{%d" s i
```
Comment (jonludlam) [inline src/model/names.ml:51]: Can we change this to `{%s}%d` please? It just looks a bit wrong as it is!

**Comment by jonludlam** [comment]
This'll need a quick rebase now the previous PR has been merged.

**Comment by lubegasimon** [comment]
> This'll need a quick rebase now the previous PR has been merged.

Done

File: src/loader/ident_env.cppo.ml [hunk 3]
```diff
@@ -152,6 +155,7 @@ let extract_extended_open o =
       extract_extended_open_items o.open_bound_items
 #endif
 
+let doc_off_mode_on = ref false
```
Comment (jonludlam) [inline src/loader/ident_env.cppo.ml:158]: I don't really like this global var - we could instead simply have this as a parameter to `extract_signature_tree_item`

Comment (jonludlam) [reply src/loader/ident_env.cppo.ml:158]: Particularly since the scope of the stop comment is only supposed to be until the end of the current module/module type/class/class type (https://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html#ss:ocamldoc-stop)

**Review by jonludlam — CHANGES_REQUESTED** [review]
Good so far, just a couple of changes needed.

File: src/compat/odoc_compat.ml [hunk 4]
```diff
@@ -2,30 +2,30 @@
...
-  include Char
-
-#if OCAML_MAJOR = 4 && OCAML_MINOR = 02
-  let lowercase_ascii = lowercase
-#else
-  let lowercase_ascii = lowercase_ascii
-#endif
-
-end
+ module String =
```
[the remark this answers is not in the document]
Comment (lubegasimon) [reply src/compat/odoc_compat.ml:5]: I think that was a result of running `dune build @fmt`, and then `dune promote`

File: src/document/generator.ml [hunk 5]
```diff
@@ -472,6 +487,7 @@ module Make (Syntax : SYNTAX) = struct
       | _ :: _ ->
           let rows =
             cstrs
+            |> List.filter (fun cstr -> not (internal_cstr_arg cstr))
```
[the remark this answers is not in the document]
Comment (lubegasimon) [reply src/document/generator.ml:490]: We remove a constructor if it has only one hidden argument, if it has more than that, then we remove the hidden argument itself.

File: src/document/generator.ml [hunk 6]
```diff
@@ -689,13 +705,15 @@ module Make (Syntax : SYNTAX) = struct
...
-            @ content
+            if List.length content > 0 then
+              O.documentedSrc
+                ( O.txt " = "
+                ++
+                if need_private then
+                  O.keyword Syntax.Type.private_keyword ++ O.txt " "
+                else O.noop )
+              @ content
+            else O.documentedSrc (O.txt "")
```
Comment (Julow) [inline src/document/generator.ml:716]:
```suggestion
            else []
```

File: src/loader/doc_attr.ml [hunk 7]
```diff
@@ -28,7 +28,7 @@ let empty : Odoc_model.Comment.docs = empty_body
 
 
 
-let load_payload : Parsetree.payload -> (string * Location.t) option = function
+let load_payload : Parsetree.payload -> string * Location.t = function
```
Comment (Julow) [inline src/loader/doc_attr.ml:31]: Why this change ? It's unrelated and move an `assert false` deeper in the code.

File: src/compat/odoc_compat.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Julow) [reply src/compat/odoc_compat.ml:5]: `dune build @fmt` shouldn't modify this file, I think you can safely revert this.

File: src/document/generator.ml [hunk 8]
```diff
@@ -455,6 +455,21 @@ module Make (Syntax : SYNTAX) = struct
...
+      | Tuple type_exprs ->
+          List.exists
+            (fun type_expr ->
+              match type_expr with
+              | Constr (p, _) ->
+                  let open Paths.Path in
+                  is_hidden (p :> Paths.Path.t)
+              | _ -> false)
+            type_exprs
+      | Record _ -> false
```
Comment (Julow) [inline src/document/generator.ml:471]: We should handle this case too:
```
type t =
  Field of { field : For_generated_code.t }
```

File: src/document/generator.ml [hunk 9]
```diff
@@ -455,6 +455,21 @@ module Make (Syntax : SYNTAX) = struct
...
+      let open Lang.TypeExpr in
+      match t.args with
+      | Tuple type_exprs ->
+          List.exists
+            (fun type_expr ->
+              match type_expr with
+              | Constr (p, _) ->
+                  let open Paths.Path in
+                  is_hidden (p :> Paths.Path.t)
+              | _ -> false)
```
Comment (Julow) [inline src/document/generator.ml:469]: What about the other cases that recursively contain an `TypeExpr.t` ? `Alias`, `Arrow`, `Tuple`, `Constr`/`Class` arguments.

For example:
```ocaml
type t = Field of (int * For_generated_code.t) * int (* Tuple *)
type t = Field of For_generated_code.t option (* Constr argument *)
```

File: src/document/generator.ml (hunk 8 above)
Comment (lubegasimon) [reply src/document/generator.ml:471]: good catch

File: src/loader/doc_attr.ml (hunk 7 above)
Comment (lubegasimon) [reply src/loader/doc_attr.ml:31]: The motive that led to this change was so as to not parse the content doc comments directly, rather check for `/*`

**Review by jonludlam — APPROVED** [review]
